mono/utils/mono-path.c

```diff
@@ -177,27 +177,137 @@ mono_path_resolve_symlinks (const char *path)
 #endif
 }
 
+static gboolean
+mono_path_char_is_separator (char ch)
+{
+#ifdef HOST_WIN32
+	return ch == '/' || ch == '\\';
+#else
+	return ch == '/';
+#endif
+}
+
+static gboolean
+mono_path_contains_separator (const char *path, size_t length)
+{
+	for (size_t i = 0; i < length; ++i) {
+		if (mono_path_char_is_separator (path [i]))
+			return TRUE;
+	}
+	return FALSE;
+}
+
+static void
+mono_path_remove_trailing_path_separators (const char *path,  size_t *length)
+{
+	size_t i = *length;
+	while (i > 0 && mono_path_char_is_separator (path [i - 1]))
+		i -= 1;
+	*length = i;
+}
+
+#ifdef HOST_WIN32
+
+static gboolean
+mono_path_char_is_lowercase (char ch)
+{
+	return ch >= 'a' && ch <= 'z';
+}
+
+// Version-specific unichar2 upcase tables are stored per-volume at NTFS format-time.
+// This is just a subset.
+static char
+mono_path_char_upcase (char a)
+{
+	return mono_path_char_is_lowercase (a) ? (char)(a - 'a' + 'A') : a;
+}
+
+static gboolean
+mono_path_char_equal (char a, char b)
+{
+	return a == b
+		|| mono_path_char_upcase (a) == mono_path_char_upcase (b)
+		|| (mono_path_char_is_separator (a) && mono_path_char_is_separator (b));
+}
+
+#endif
+
+static gboolean
+mono_path_equal (const char *a, const char *b, size_t length)
+{
+#ifdef HOST_WIN32
+	size_t i = 0;
+	for (i = 0; i < length && mono_path_char_equal (a [i], b [i]); ++i) {
+		// nothing
+	}
+	return i == length;
```

Comment:
Consider return false in the loop, true after it. Clarity of current form is debatable. It is an idiom but maybe a stretch for too many people and not worth while. One instance, no big deal. More important is widespread convention.